### ISSUE TITLE
Squiz.Arrays.ArrayDeclaration with tabs

### ISSUE THREAD

**Issue by josephzidell** [squizlabs/PHP_CodeSniffer#1795]
Seems `Squiz.Arrays.ArrayDeclaration` doesn't play well with tab indentation.

v 3.2.0

**Comment by josephzidell** [comment]
Example:
```php
<?php
$arr = [
	'foo',
	'bar',
];
```

Gives these errors:
``` 3 | ERROR | [x] Array value not aligned correctly; expected 8 spaces but found 1
 4 | ERROR | [x] Array value not aligned correctly; expected 8 spaces but found 1
 5 | ERROR | [x] Closing parenthesis not aligned correctly; expected 7 space(s) but found 0
```

**Comment by gsherwood** [comment]
Did you specify a tab width on the command line (`--tab-width=4`)? That tells PHPCS to replace tabs with spaces, allowing sniffs that don't do specific tab checks (pretty much all of them) to check indents correctly.

You can also put this directly into a custom coding standard using:
```xml
<arg name="tab-width" value="4"/>
```

**Comment by josephzidell** [comment]
Unfortunately, that made is slightly worse.

Here's my code:
```php
class PurchaseOrder
{
	public $hasMany = [
		'PurchaseOrderVendorAddress' => [
			'something',
		],
	];
}
```

Without the setting, it gives
```Array key not aligned correctly; expected 20 spaces but found 2```
With the setting, it gives
`Array key not aligned correctly; expected 23 spaces but found 8`

**Comment by jrfnl** [comment]
@josephzidell Just wondering - could it be that you want/expect "normalized" array indents ? I.e. closer aligned with start of array opener and items indented one tab-size in ? In that case, you may want to check out the new `Generic.Arrays.ArrayIndent` sniff.
The `Squiz` version expects the closer to be exactly aligned with the opener (note: not the line containing the opener, the opener itself!) and the items to be indented from there.

**Comment by gsherwood** [comment]
Exactly what @jrfnl said. You probably don't want the array formatting rules of the Squiz standard. Try the Generic sniff instead.

**Comment by josephzidell** [comment]
OK. I tried that one, and it works for basic indentation, not rockets or values.
```php
	'className' => 'Customer',
	'foreignKey' => 'entity_id',
```
After reading the release notes for 3.2.0, I thought there was a sniff to detect alignment of arrows and values. I may have been wrong.

**Comment by gsherwood** [comment]
> After reading the release notes for 3.2.0, I thought there was a sniff to detect alignment of arrows and values. I may have been wrong.

You may be referring to the AbstractArraySniff, which is an abstract class that can be used to build custom array sniffs. The Generic.Arrays.ArrayIndent sniff does that, but it only does basic indents checks. There is no generic sniff to check for alignment of arrays. The Squiz sniff does that amongst other checks, but you don't want your array formatted like that.

**Comment by josephzidell** [comment]
I want it thus:
```php
	'className'  => 'Customer',
	'foreignKey' => 'entity_id',
```

**Comment by josephzidell** [comment]
Main thing is to line up the arrows. Can we do that?

**Comment by jrfnl** [comment]
@gsherwood I have a sniff for that which I wrote for WPCS. If you're open to it, I could adjust it to use the new abstract base sniff and pull it here ?

**Comment by gsherwood** [comment]
@jrfnl I plan of building a sniff to just do arrow alignment eventually, so if you have one that just does that, I'd be very happy to take a look.

**Comment by jrfnl** [comment]
@gsherwood I have one which does just that - including some configurable settings -. I also have a separate one which checks the comma's (none after last item in single line, comma after last item in multi-line).
Both were written to get round various bugs in the Squiz array sniff when parts of that were disabled, as well as to fix bugs with nowdoc/heredoc syntaxes.

Both are based on a set of utility functions which are used in both WPCS as well as PHPCompatibility and which I intended to pull here at some point as well. I expect the utility functions to have significant overlap with your abstract array sniff, though I'll have to check that in more detail.
The principle approach of the utility functions is slightly different though as they will also work for function call parameter splitting.
I waited to pull these as I wanted to make sure they were as bug-free as possible, which by now I'm pretty sure they are as they've been used for over a year.

You can find the code I'm talking about here:
* Utility functions: https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/blob/cc7d15d9de07502ead2c4515dd0c3007d62dacaf/WordPress/Sniff.php#L1919-L2181 and the PHPCompatibility standard also has dedicated unit tests for them: 
    - https://github.com/wimg/PHPCompatibility/blob/master/PHPCompatibility/Tests/BaseClass/GetFunctionParametersTest.php
    - https://github.com/wimg/PHPCompatibility/blob/master/PHPCompatibility/Tests/BaseClass/GetFunctionParameterCountTest.php
    - https://github.com/wimg/PHPCompatibility/blob/master/PHPCompatibility/Tests/BaseClass/DoesFunctionCallHaveParametersTest.php
* Array arrow alignment sniff: https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/blob/develop/WordPress/Sniffs/Arrays/MultipleStatementAlignmentSniff.php
* Array comma sniff: https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/blob/develop/WordPress/Sniffs/Arrays/CommaAfterArrayItemSniff.php

**Comment by gsherwood** [comment]
I don't think I'd want to support those options in the array arrow alignment sniff. You should probably keep that one.

I'm actually going to close this issue because the sniff is working as expected and there are already other issues related to creating new array sniffs in this project. I think I'll probably just continue with the plan of looking at that in a future version.

**Comment by jrfnl** [comment]
> I don't think I'd want to support those options in the array arrow alignment sniff.

@gsherwood Just so you know: those options basically mirror the options available in the generic assignment operator (`=` and the likes) alignment sniff.
